### PR TITLE
Added default role tab to realm settings

### PR DESCRIPTION
**Pull request by edewit**
Also changed that the default role in realm tab now navigates to the user registration tab off the realm settings.
And introduced a `KeycloakSpinner` to remove some duplicate code.

### CODE REVIEW

**Review by jenny-s51 — DISMISSED** [review]
Great work @edewit ! Associated role inheritance works as expected both in `Realm roles -> associated roles` and `Realm settings -> User settings -> default role`.

I noticed the pagination isn't updating the data correctly, though after some investigating it seems like this can be resolved by simply removing the `isPaginated` prop in the KDT of the `AssociatedRolesTab`.

How would you suggest testing your changes in client-scopes? I was able to view the `AddRoleMappingModal`  by selecting a hardcoded role mapper type for an LDAP provider, but under client scopes I don't see any of these mapping types: `"service-account" | "client-scope" | "role"`.

File: src/realm-settings/UserRegistration.tsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import { useRealm } from "../context/realm-context/RealmContext";
+import { AssociatedRolesTab } from "../realm-roles/AssociatedRolesTab";
+import { KeycloakSpinner } from "../components/keycloak-spinner/KeycloakSpinner";
+
+type Realm = RealmRepresentation & {
+  defaultRole: {
+    id: string;
+    name: string;
+  };
+};
```
Comment (jonkoops) [inline src/realm-settings/UserRegistration.tsx:16]: Why do we need this type? Is the type from the admin client incorrect? Can we fix this there as well?

File: src/realm-roles/routes/ClientRole.ts [hunk 2]
```diff
@@ -7,7 +7,7 @@ export type ClientRoleTab =
   | "details"
   | "attributes"
   | "users-in-role"
-  | "AssociateRoles";
+  | "AssociatedRoles";
```
Comment (jonkoops) [inline src/realm-roles/routes/ClientRole.ts:10]: The casing of this key is different, perhaps we can align it with the rest?

File: src/realm-roles/RolesList.tsx [hunk 3]
```diff
@@ -79,21 +81,25 @@ export const RolesList = ({
...
-  const RoleDetailLink = (role: RoleRepresentation) => (
-    <>
+  const RoleDetailLink = (role: RoleRepresentation) =>
+    role.name !== realm?.defaultRole?.name ? (
       <RoleLink role={role}>{role.name}</RoleLink>
-      {role.name?.includes("default-role") ? (
+    ) : (
+      <>
+        <Link
+          key={role.id}
```
Comment (jonkoops) [inline src/realm-roles/RolesList.tsx:90]: I don't think we need a key here, unless I am missing some iteration.

File: src/realm-roles/RealmRoleTabs.tsx [hunk 4]
```diff
@@ -284,14 +290,50 @@ export default function RealmRoleTabs() {
...
+      : toRealmRole({
+          realm: realm?.realm!,
+          id,
+          tab: "AssociatedRoles",
+        });
+    history.push(to);
+  };
+
+  const addComposites = async (composites: RoleRepresentation[]) => {
+    const compositeArray = composites;
```
Comment (jonkoops) [inline src/realm-roles/RealmRoleTabs.tsx:311]: This assignment seems pretty useless as it creates an alias which is only passed once.

**Review by jonkoops — APPROVED** [review]

File: src/realm-settings/UserRegistration.tsx (hunk 1 above)
Comment (edewit) [reply src/realm-settings/UserRegistration.tsx:16]: It's already fixed in the new adminClient

File: src/realm-roles/routes/ClientRole.ts (hunk 2 above)
Comment (edewit) [reply src/realm-roles/routes/ClientRole.ts:10]: it's was already like this... I can change it

File: src/realm-roles/RolesList.tsx (hunk 3 above)
Comment (edewit) [reply src/realm-roles/RolesList.tsx:90]: right

File: src/realm-settings/RealmSettingsTabs.tsx [hunk 5]
```diff
@@ -432,6 +433,14 @@ export const RealmSettingsTabs = ({
                 </Tab>
               </Tabs>
             </Tab>
+            <Tab
+              eventKey="userRegistration"
+              title={<TabTitleText>{t("userRegistration")}</TabTitleText>}
+              data-testid="rs-clientPolicies-tab"
+              aria-label={t("clientPoliciesTab")}
```
[the remark this answers is not in the document]
Comment (edewit) [reply src/realm-settings/RealmSettingsTabs.tsx:440]: @jenny-s51 just removing the`isPaginated` can't be the solution right as it seems that `first` and `max` are used to find composite roles

[the remark this answers is not in the document]
Comment (edewit) [reply src/realm-settings/RealmSettingsTabs.tsx]: > How would you suggest testing your changes in client-scopes? I was able to view the AddRoleMappingModal by selecting a hardcoded role mapper type for an LDAP provider, but under client scopes I don't see any of these mapping types: "service-account" | "client-scope" | "role".

I didn't change anything on client scopes only introduced this new screen

File: src/realm-settings/UserRegistration.tsx (hunk 1 above)
Comment (edewit) [reply src/realm-settings/UserRegistration.tsx:16]: can be removed when #1522 is merged

**Review by jenny-s51 — COMMENTED** [review]
Pagination looks great and everything else works as expected @edewit! Using `isInheritedHidden` on the isPaginated prop was a good idea.

File: src/realm-roles/routes/ClientRole.ts [hunk 6]
```diff
@@ -7,7 +7,7 @@ export type ClientRoleTab =
   | "details"
   | "attributes"
   | "users-in-role"
-  | "AssociateRoles";
+  | "associated-roles";
```
Comment (jenny-s51) [inline src/realm-roles/routes/ClientRole.ts:10]: Looks like the Cypress test just needs to be updated to reflect this new route

File: src/realm-settings/RealmSettingsTabs.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (jenny-s51) [reply src/realm-settings/RealmSettingsTabs.tsx]: Ah I see, yes that makes sense. Thank you for clarifying

File: src/realm-settings/UserRegistration.tsx (hunk 1 above)
Comment (jonkoops) [reply src/realm-settings/UserRegistration.tsx:16]: @edewit This seems to be merged now, can we consider this resolved?

**Review by jonkoops — APPROVED** [review]

**Comment by edewit** [comment]
@jenny-s51 I don't see the client tests fail?

File: src/realm-roles/routes/ClientRole.ts (hunk 6 above)
Comment (edewit) [reply src/realm-roles/routes/ClientRole.ts:10]: I don't see the client tests fail?

File: src/realm-roles/RolesList.tsx [hunk 7]
```diff
@@ -72,21 +70,24 @@ export const RolesList = ({
...
+        <Link
+          to={toRealmSettings({ realm: realmName, tab: "userRegistration" })}
+        >
+          {role.name}{" "}
+        </Link>
         <HelpItem
           helpText={t("defaultRole")}
           forLabel={t("defaultRole")}
           forID="kc-defaultRole"
           id="default-role-help-icon"
```
Comment (agagancarczyk) [inline src/realm-roles/RolesList.tsx:87]: is this id needed? we don't seem to use it in some other components

File: src/realm-roles/RolesList.tsx [hunk 8]
```diff
@@ -72,21 +70,24 @@ export const RolesList = ({
...
+      <>
+        <Link
+          to={toRealmSettings({ realm: realmName, tab: "userRegistration" })}
+        >
+          {role.name}{" "}
+        </Link>
         <HelpItem
           helpText={t("defaultRole")}
           forLabel={t("defaultRole")}
           forID="kc-defaultRole"
```
Comment (agagancarczyk) [inline src/realm-roles/RolesList.tsx:86]: for accessibility, we use this format for forID={t(`common:helpLabel`, {
                    label: t("defaultRole"),
                  })}

File: src/realm-roles/RolesList.tsx (hunk 7 above)
Comment (edewit) [reply src/realm-roles/RolesList.tsx:87]: it's used in the css

File: src/realm-roles/routes/ClientRole.ts (hunk 6 above)
Comment (jenny-s51) [reply src/realm-roles/routes/ClientRole.ts:10]: It looks like we can get the `realm_roles` test to pass by updating this line with `/associated-roles` https://github.com/keycloak/keycloak-admin-ui/blob/main/cypress/support/pages/admin_console/manage/realm_roles/AssociatedRolesPage.ts#L21

**Review by agagancarczyk — APPROVED** [review]

**Review by jonkoops — APPROVED** [review]

**Pull request merged**